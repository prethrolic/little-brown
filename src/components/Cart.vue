<template>
  <div class='cart-container'>
    <div class='cart-header-container'>
      <h2 class='cart-title'>Cart</h2>
      <div class='cart-clear-button' @click='onClear'>
        Clear All
      </div>
    </div>
    <div class='cart-products-container'>
      <CartItem
        v-for='id in cartItemList'
        :key='id'
        :book='cart[id].book'
        :count='cart[id].count'
      />
    </div>
    <div class='cart-total-price-container'>
      <table>
        <tr>
          <td class='cart-small-label'>Subtotal</td>
          <td v-html="'฿' + subtotal" class='cart-small-value' />
        </tr>
        <tr>
          <td class='cart-small-label'>Discounts</td>
          <td v-html="'- ฿' + discount" class='cart-small-value'/>
        </tr>
        <tr>
          <td class='cart-small-label'>Sales Tax</td>
          <td v-html="'฿' + tax.toFixed(2)" class='cart-small-value' />
        </tr>
      </table>
      <div class='cart-divider' />
      <table>
        <tr>
          <td class='cart-large-label'>Total</td>
          <td v-html="'฿' + total" class='cart-large-value' />
        </tr>
      </table>
    </div>
    <div
      :class="['cart-payment-button', { 'disabled' : isCartEmpty }]"
      @click='onPayment'
    >
      Pay with Cash
    </div>
    <PaymentModal
      :total='total'
      :isOpen='isPaymentModalOpen'
      @payment:cancel='isPaymentModalOpen = false'
      @payment:success='onPaymentSuccess'
    />
    <ReceiptModal
      :date='new Date()'
      :cart='cart'
      :user='user'
      :paymentInfo='{ subtotal, discount, total, cash }'
      :isOpen='isReceiptModalOpen'
      @bill:cancel='isReceiptModalOpen = false'
      @bill:success='onBillSuccess'
    />
  </div>
</template>
<script>
import CartItem from '@/components/CartItem.vue';
import PaymentModal from '@/components/PaymentModal.vue';
import ReceiptModal from '@/components/ReceiptModal.vue';

export default {
  name: 'Cart',
  components: {
    CartItem,
    PaymentModal,
    ReceiptModal,
  },
  data() {
    return {
      cart: {},
      cartItemList: [],
      inPromotion: [],
      tax: 0,
      cash: '',
      isPaymentModalOpen: false,
      isReceiptModalOpen: false,
      scroll: false,
    };
  },
  updated() {
    if (this.scroll) {
      this.scroll = false;
      this.$nextTick(() => this.scrollToBottom());
    }
  },
  methods: {
    getCart() {
      this.cart = this.$store.state.cart;
      this.cartItemList = Object.keys(this.cart);
      this.inPromotion = this.$store.state.inPromotion;
    },
    onBillSuccess() {
      this.$store.dispatch('clear_all');
      window.location.reload();
    },
    onClear() {
      this.$store.dispatch('clear_all');
    },
    onPayment() {
      this.isPaymentModalOpen = true;
    },
    onPaymentSuccess(cash) {
      this.isPaymentModalOpen = false;
      this.cash = cash;
      this.isReceiptModalOpen = true;
    },
    scrollToBottom() {
      const productContainer = this.$el.querySelector('.cart-products-container');
      productContainer.scrollTop = productContainer.scrollHeight;
    },
  },
  computed: {
    discount() {
      let discount = 0;
      const promotedTitleCount = this.inPromotion.length;
      if (promotedTitleCount > 1 && promotedTitleCount <= 7) {
        this.inPromotion.forEach((id) => {
          discount += this.cart[id].book.price * (10 + (promotedTitleCount - 2)) * 0.01;
        });
      } else if (promotedTitleCount > 7) {
        this.inPromotion.forEach((id) => {
          discount += this.cart[id].book.price * 0.15;
        });
      }

      return discount.toFixed(2);
    },
    isCartEmpty() {
      return this.cartItemList.length === 0;
    },
    total() {
      return (this.subtotal - this.discount + this.tax).toFixed(2);
    },
    subtotal() {
      let subtotal = 0;
      this.cartItemList.forEach((id) => {
        subtotal += this.cart[id].book.price * this.cart[id].count;
      });

      return subtotal.toFixed(2);
    },
    user() {
      return this.$store.state.user;
    },
  },
  watch: {
    '$store.state.cart': {
      deep: true,
      handler(oldCart, newCart) {
        this.getCart();
        if (Object.keys(oldCart).length > Object.keys(newCart).length) {
          this.scroll = true;
        }
      },
    },
  },
  mounted() {
    this.getCart();
  },
};
</script>
<style lang="scss" scoped>
@import '@/assets/scss/variables.scss';

.cart {
  &-container {
    flex: 0 0 30%;
    display: flex;
    flex-direction: column;
    padding: 24px;
    height: 100vh;
    overflow-y: scroll;
    border-right: 1px solid $light-gray;
    background-color: $bg-gray;
  }

  &-header-container {
    display: flex;
    margin: 24px 0;
  }

  &-clear-button {
    margin-left: auto;
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 16px;
    background-color: $secondary;
    border-radius: 12px;
    font-size: 12px;
    font-weight: bold;
    color: white;
    cursor: pointer;
    transition: 0.2s ease-in-out;

    &:hover {
      background-color: $secondary-hover;
      color: white;
      cursor: pointer;
    }
  }

  &-products-container {
    height: 100%;
    min-height: 200px;
    margin-top: 16px;
    overflow-y: scroll;
  }

  &-total-price-container {
    width: 100%;
    background-color: white;
    border: 1px solid $light-gray;
    border-radius: 12px;
    padding: 8px 16px;
    margin-top: 24px;
    -webkit-box-shadow: 0px 0px 21px -5px rgba(199,199,199,0.44);
    box-shadow: 0px 0px 21px -5px rgba(199,199,199,0.44);
  }

  &-small-label {
    font-size: 14px;
    text-align: left;
  }

  &-small-value {
    font-size: 14px;
    font-weight: bold;
    text-align: right;
  }

  &-large-label {
    font-size: 18px;
    text-align: left;
  }

  &-large-value {
    font-size: 20px;
    font-weight: bold;
    text-align: right;
  }

  &-divider {
    background-color: $light-gray;
    height: 1px;
    width: 100%;
    margin: 4px 0;
  }

  &-payment-button {
    display: flex;
    align-items: center;
    justify-content: center;
    margin: 24px 0;
    width: 100%;
    min-height: 64px;
    border-radius: 12px;
    background-color: $primary;
    font-size: 20px;
    font-weight: bold;
    color: white;
    transition: 0.2s ease-in-out;
    cursor: pointer;

    &:hover {
      background-color: $primary-hover;
    }

    &.disabled {
      background-color: $light-gray;
      color: $dark-gray;
      cursor: not-allowed;
    }
  }
}

table {
  width: 100%;
  padding: 8px 8px 4px 8px;
}

td {
  padding-bottom: 8px;
}

@media(max-width: 1112px) {
  .cart {
    &-container {
      flex: 0 0 35%;
    }
  }
}

@media(max-width: 1024px) {
  .cart {
    &-container {
      flex: 0 0 60%;
      padding: 24px 36px;
    }
  }
}

@media(max-width: 768px) {
  .cart {
    &-container {
      flex: 0 0 70%;
      padding: 24px 36px;
    }

    &-payment-button {
      margin: 16px 0;
      min-height: 48px;
      font-size: 14px;
    }

    &-small-label, &-small-value  {
      font-size: 12px;
    }

    &-large-label {
      font-size: 14px;
      text-align: left;
    }

    &-large-value {
      font-size: 16px;
    }
  }
}

@media(max-width: 600px) {
  .cart {
    &-container {
      flex: 0 0 85%;
      padding: 24px;
    }
  }
}
</style>
