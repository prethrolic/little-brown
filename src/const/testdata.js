export const bookList = [
  {
    "cover": "https://d1w7fb2mkkr3kw.cloudfront.net/assets/images/book/mid/9781/4088/9781408855652.jpg",
    "price": "350",
    "title": "Harry Potter and the Philosopher's Stone (I)",
    "id": "9781408855652"
  },
  {
    "cover": "https://d1w7fb2mkkr3kw.cloudfront.net/assets/images/book/mid/9781/4088/9781408855669.jpg",
    "price": "350",
    "title": "Harry Potter and the Chamber of Secrets (II)",
    "id": "9781408855669"
  },
  {
    "cover": "https://d1w7fb2mkkr3kw.cloudfront.net/assets/images/book/mid/9781/4088/9781408855676.jpg",
    "price": "340",
    "title": "Harry Potter and the Prisoner of Azkaban (III)",
    "id": "9781408855676"
  },
  {
    "cover": "https://d1w7fb2mkkr3kw.cloudfront.net/assets/images/book/mid/9780/2419/9780241975367.jpg",
    "price": "230",
    "title": "The Monk of Mokha",
    "id": "9780241975367"
  },
  {
    "cover": "https://d1w7fb2mkkr3kw.cloudfront.net/assets/images/book/mid/9781/7847/9781784706883.jpg",
    "price": "300",
    "title": "The Language of Kindness",
    "id": "9781784706883"
  },
];

export const cart = {
  "9781408855652": {
    book: {
    "cover": "https://d1w7fb2mkkr3kw.cloudfront.net/assets/images/book/mid/9781/4088/9781408855652.jpg",
    "price": "350",
    "title": "Harry Potter and the Philosopher's Stone (I)",
    "id": "9781408855652"
    },
    count: 2,
  },
  "9781408855669": {
    book: {
      "cover": "https://d1w7fb2mkkr3kw.cloudfront.net/assets/images/book/mid/9781/4088/9781408855669.jpg",
      "price": "350",
      "title": "Harry Potter and the Chamber of Secrets (II)",
      "id": "9781408855669"
    },
    count: 1,
  },
  "9781408855676": {
    book: {
      "cover": "https://d1w7fb2mkkr3kw.cloudfront.net/assets/images/book/mid/9781/4088/9781408855676.jpg",
      "price": "340",
      "title": "Harry Potter and the Prisoner of Azkaban (III)",
      "id": "9781408855676"
    },
    count: 1,
  },
  "9780241975367": {
    book: {
      "cover": "https://d1w7fb2mkkr3kw.cloudfront.net/assets/images/book/mid/9780/2419/9780241975367.jpg",
      "price": "230",
      "title": "The Monk of Mokha",
      "id": "9780241975367"
    },
    count: 2,
  },
};

export const inPromotion = ["9781408855652", "9781408855669", "9781408855676"];

export const user = {
  displayName: 'John Doe',
  pictureUrl: 'https://pbs.twimg.com/profile_images/1269585418107805697/6QKUkoWJ_400x400.jpg',
};

export const paymentInfo = {
  subtotal: '1850.00',
  discount: '114.40',
  total: '1735.60',
  cash: '2000',
};