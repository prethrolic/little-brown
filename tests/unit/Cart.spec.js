import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import Cart from '@/components/Cart.vue';
import { cart, inPromotion } from '@/const/testdata.js';

const localVue = createLocalVue();
localVue.use(Vuex);
const originalDate = global.Date;
const mockDate = new Date(1466424490000);

describe('Cart.vue', () => {
  beforeEach(() => {
    global.Date = jest.fn().mockImplementation(() => mockDate);
  });

  afterEach(() => { 
    global.Date.mockClear();
    global.Date = originalDate; 
  });

  it('render disbaled payment button when cart is empty', async () => {
    const wrapper = shallowMount(Cart, {
      mocks: {
        $store: {
          state: {
            cart: {},
            inPromotion: [],
          },
        },
      },
      localVue
    });
    await wrapper.setData({
      cart: {},
      cartItemList: [],
    });
    expect(wrapper.vm.$el.querySelectorAll('.cart-payment-button.disabled').length).toBe(1);
    expect(wrapper.element).toMatchSnapshot();

  });

  it('render cart correctly', async () => {
    const wrapper = shallowMount(Cart, {
      mocks: {
        $store: {
          state: {
            cart,
            inPromotion,
          },
        },
      },
      localVue
    });
    await wrapper.setData({
      cart,
      cartItemList: Object.keys(cart)
    })

    expect(wrapper.element).toMatchSnapshot();
  });

  it('calculate discount correctly', async () => {
    const wrapper = shallowMount(Cart, {
      mocks: {
        $store: {
          state: {
            cart,
            inPromotion,
          },
        },
      },
      localVue
    });
    await wrapper.setData({
      cart,
      cartItemList: Object.keys(cart)
    })

    expect(wrapper.vm.discount).toEqual('114.40');
  });

  it('calculate subtotal correctly', async () => {
    const wrapper = shallowMount(Cart, {
      mocks: {
        $store: {
          state: {
            cart,
            inPromotion,
          },
        },
      },
      localVue
    });
    await wrapper.setData({
      cart,
      cartItemList: Object.keys(cart)
    })

    expect(wrapper.vm.subtotal).toEqual('1850.00');
  });

  it('calculate total correctly', async () => {
    const wrapper = shallowMount(Cart, {
      mocks: {
        $store: {
          state: {
            cart,
            inPromotion,
          },
        },
      },
      localVue
    });
    await wrapper.setData({
      cart,
      cartItemList: Object.keys(cart)
    })

    expect(wrapper.vm.total).toEqual('1735.60');
  });
});