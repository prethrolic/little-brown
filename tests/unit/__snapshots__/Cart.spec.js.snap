// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart.vue render cart correctly 1`] = `
<div
  class="cart-container"
>
  <div
    class="cart-header-container"
  >
    <h2
      class="cart-title"
    >
      Cart
    </h2>
     
    <div
      class="cart-clear-button"
    >
      
      Clear All
    
    </div>
  </div>
   
  <div
    class="cart-products-container"
  >
    <cartitem-stub
      book="[object Object]"
      count="2"
    />
    <cartitem-stub
      book="[object Object]"
      count="1"
    />
    <cartitem-stub
      book="[object Object]"
      count="1"
    />
    <cartitem-stub
      book="[object Object]"
      count="2"
    />
  </div>
   
  <div
    class="cart-total-price-container"
  >
    <table>
      <tr>
        <td
          class="cart-small-label"
        >
          Subtotal
        </td>
         
        <td
          class="cart-small-value"
        >
          ฿1850.00
        </td>
      </tr>
       
      <tr>
        <td
          class="cart-small-label"
        >
          Discounts
        </td>
         
        <td
          class="cart-small-value"
        >
          - ฿114.40
        </td>
      </tr>
       
      <tr>
        <td
          class="cart-small-label"
        >
          Sales Tax
        </td>
         
        <td
          class="cart-small-value"
        >
          ฿0.00
        </td>
      </tr>
    </table>
     
    <div
      class="cart-divider"
    />
     
    <table>
      <tr>
        <td
          class="cart-large-label"
        >
          Total
        </td>
         
        <td
          class="cart-large-value"
        >
          ฿1735.60
        </td>
      </tr>
    </table>
  </div>
   
  <div
    class="cart-payment-button"
  >
    
    Pay with Cash
  
  </div>
</div>
`;
